proof-generator/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+}
+
+/// Sign a message with a ETH private key using either a keystore file or a Ledger HW
+pub async fn eth_sign<S: 'static + Signer>(
+    signer: &S,
+    proof: &str,
+) -> Result<Signature, <S as Signer>::Error> {
+    signer.sign_message(proof).await
+}
+
+/// Create the proof to be signed
+fn create_proof<T: Display, K: Debug>(
+    ownership: &T,
+    evidence: &K,
+    block_hash: &H256,
+) -> Result<String, Error> {
+    Ok(format!(
+        "As the owner of GPG key {}, my Ethereum address is {:?} as of {:?}",
+        &ownership, &evidence, &block_hash
+    ))
+}
+
+/// Sign the created proof with GPG and ETH keypairs
+async fn sign_proof<T: 'static + Signer>(
+    gpg_key: &str,
+    signer: &T,
+    block_hash: &H256,
+) -> anyhow::Result<Proof> {
+    let msg = create_proof(&gpg_key, &signer.address(), &block_hash)?;
+
+    log::info!("Signing message with ETH keypair..");
+    let eth_sig = eth_sign(signer, &msg)
+        .await
+        .map_err(|_| anyhow!(Error::ETHSigFailed))?;
+    log::debug!("ETH Signature: {:?}.", eth_sig);
+
+    log::info!("Signing message with GPG keypair..");
+    let gpg_sig = gpg_sign(&gpg_key, &msg)?;
+    log::debug!("GPG Signature: {:?}.", gpg_sig);
+
+    Ok(Proof { gpg_sig, eth_sig })
+}
+
+/// The main lib function that runs the functionality of the program
+/// - Obtains a block hash from a block from 1 day ago.
+/// - Gets either a keystore file or in its absence a Ledger HW as signer to sign a message.
+/// - Creates a message that will be signed by the defined signer.
+/// - Write both proofs to a JSON file.
+pub async fn run(opts: Options) -> anyhow::Result<()> {
+    let provider =
+        Provider::<Http>::try_from(opts.rpc_url).expect("could not instantiate HTTP Provider");
+    let latest_block_number = provider.get_block_number().await?;
+    // 5760 blocks earlier is aprox. 1 day ago, this due to avoid referencing blocks that are affected by reorgs of the chain.
+    let block_number = latest_block_number - 5760u64;
+    if let Some(block) = provider.get_block(block_number).await? {
+        // TODO: This is redundant eventually, since when a block is found by ethers-rs it must have a hash... I kept it to maintain the code style
+        if let Some(block_hash) = block.hash {
+            if let Some(keypath) = &opts.keystore {
+                use colored::*;
+
+                log::info!("Decrypting keystore..");
+                let prompt = format!("{} Password: ", "??".cyan());
+                let password = rpassword::prompt_password_stdout(&prompt).unwrap();
+                let signer = ethers::signers::LocalWallet::decrypt_keystore(keypath, password)
+                    .map_err(|_| anyhow!("keystore decryption failed"))?;
+                log::debug!("Keystore decrypted: {:?}.", signer);
+
+                let proof = sign_proof(&opts.gpg_key, &signer, &block_hash).await?;
+                fs::write(&opts.output, serde_json::to_string(&proof)?)?;
+                Ok(())
```

Comment:
Stylistic thing, but in big blocks of code, I like to keep the last statement separated by a newline to make the code easier to scan, eg.:

```
foo
bar
baz

Ok(())
```

```suggestion

                Ok(())
```